File: common/src/main/java/wile/redstonepen/blocks/controlbox/ControlBoxBlockEntity.java
package wile.redstonepen.blocks.controlbox;

import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.core.HolderLookup;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.Tag;
import net.minecraft.network.chat.Component;
import net.minecraft.util.Mth;
import net.minecraft.world.Container;
import net.minecraft.world.MenuProvider;
import net.minecraft.world.Nameable;
import net.minecraft.world.SimpleContainer;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.inventory.ContainerLevelAccess;
import net.minecraft.world.inventory.SimpleContainerData;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.state.BlockState;
import wile.redstonepen.detail.RcaSync;
import wile.redstonepen.util.Auxiliaries;
import wile.redstonepen.net.Networking;
import wile.redstonepen.registry.Registries;
import wile.redstonepen.blocks.StandardEntityBlocks;

import org.jetbrains.annotations.Nullable;
import java.util.Map;
import java.util.UUID;


@SuppressWarnings("deprecation")
public class ControlBoxBlockEntity extends StandardEntityBlocks.StandardBlockEntity implements MenuProvider, Nameable, Networking.IPacketTileNotifyReceiver
{
  public static final int TICK_INTERVAL = 4;
  private final Container block_inventory_ = new SimpleContainer(1);
  private final ControlBoxLogic logic_ = new ControlBoxLogic();
  private UUID activating_player_ = null;
  private Component custom_name_ = null;
  private boolean trace_ = false;
  private int tick_timer_ = 0;
  private int tick_interval_ = 0;

  public ControlBoxBlockEntity(BlockPos pos, BlockState state)
  { super(Registries.getBlockEntityTypeOfBlock(state.getBlock()), pos, state); }

  @Override
  public CompoundTag readnbt(HolderLookup.Provider hlp, CompoundTag nbt)
  {
    if(nbt.contains("name", Tag.TAG_STRING)) custom_name_ = Auxiliaries.unserializeTextComponent(nbt.getString("name"), hlp);
    final CompoundTag logic_data = nbt.contains("logic", Tag.TAG_COMPOUND) ? nbt.getCompound("logic") : new CompoundTag();
    logic_.code(logic_data.getString("code"));
    logic_.input_data = logic_data.getInt("input");
    logic_.output_data = logic_data.getInt("output");
    final CompoundTag logic_symbols = logic_data.contains("symbols", Tag.TAG_COMPOUND) ? logic_data.getCompound("symbols") : new CompoundTag();
    logic_.clearSymbols();
    logic_symbols.getAllKeys().forEach(k->logic_.symbol(k, logic_symbols.getInt(k)));
    activating_player_ = nbt.hasUUID("player") ? nbt.getUUID("player") : null;
    return nbt;
  }

  @Override
  public CompoundTag writenbt(HolderLookup.Provider hlp, CompoundTag nbt, boolean sync_packet)
  {
    if(custom_name_ != null) nbt.putString("name", Auxiliaries.serializeTextComponent(custom_name_, hlp));
    final CompoundTag logic_data = new CompoundTag();
    logic_data.putString("code", logic_.code());
    logic_data.putInt("input", logic_.input_data);
    logic_data.putInt("output", logic_.output_data);
    final CompoundTag logic_symbols = new CompoundTag();
    logic_.symbols().forEach(logic_symbols::putInt);
    logic_data.put("symbols", logic_symbols);
    nbt.put("logic", logic_data);
    if(activating_player_ != null) nbt.putUUID("player", activating_player_);
    return nbt;
  }

  // BlockEntity/MenuProvider -------------------------------------------------------

  @Override
  public Component getName()
  {
    if(custom_name_ != null) return custom_name_;
    return Component.translatable(getBlockState().getBlock().getDescriptionId());
  }

  @Override
  @Nullable
  public Component getCustomName()
  { return custom_name_; }

  @Override
  public boolean hasCustomName()
  { return (custom_name_ != null); }

  public void setCustomName(Component name)
  { custom_name_ = name; }

  @Override
  public Component getDisplayName()
  { return Nameable.super.getDisplayName(); }

  @Override
  public AbstractContainerMenu createMenu(int id, Inventory inventory, Player player)
  { return new ControlBoxUiContainer(id, inventory, block_inventory_, ContainerLevelAccess.create(level, worldPosition), new SimpleContainerData(1)); }

  @Override
  public void tick()
  {
    if(--tick_timer_ > 0) return;
    tick_timer_ = (tick_interval_>0) ? tick_interval_ : TICK_INTERVAL;
    final long tick = System.nanoTime();
    final Level world = getLevel();
    final BlockState device_state = getBlockState();
    final BlockPos device_pos = getBlockPos();
    final boolean device_enabled = (device_state.getValue(ControlBoxBlock.STATE) > 0) || (device_state.getValue(ControlBoxBlock.POWERED));
    if(!(device_state.getBlock() instanceof final ControlBoxBlock device_block)) return;
    final int last_output_data = logic_.output_data;
    final int last_input_data = logic_.input_data;
    final RcaSync.RcaData rca_data = (((logic_.rca_input_mask|logic_.rca_output_mask)==0) ? (RcaSync.CommonRca.EMPTY) : RcaSync.CommonRca.ofPlayer(activating_player_, false));
    try {
      // Input fetching
      {
        logic_.input_data = 0;
        for(Direction d:Direction.values()) {
          final Direction world_dir = ControlBoxBlock.getForwardStateMappedFacing(device_state, d);
          if(device_enabled) {
            // Comparator overrides only if really needed - may be inventories that do expensive lookups.
            final String port_name = Defs.PORT_NAMES.get(d.ordinal());
            if(logic_.usesSymbol(port_name+".co")) {
              final BlockPos target_pos = device_pos.relative(world_dir);
              final net.minecraft.world.level.block.state.BlockState target_state = world.getBlockState(target_pos);
              if(target_state.hasAnalogOutputSignal()) {
                final int cov = target_state.getAnalogOutputSignal(world, target_pos);
                logic_.symbol(port_name+".co", cov);
              } else {
                logic_.symbol(port_name+".co", 0);
              }
            }
          }
          if((logic_.output_mask & (0xf<<(4*d.ordinal()))) == 0) {
            final int p = world.getSignal(device_pos.relative(world_dir), world_dir);
            logic_.input_data |= (p & 0xf)<<(4*d.ordinal());
          }
        }
        if((logic_.rca_input_mask != 0) && (rca_data != RcaSync.CommonRca.EMPTY)) logic_.rca_input_data = (rca_data.client_inputs() & logic_.rca_input_mask);
      }
      // Logic processing
      {
        if(trace_) logic_.symbol(".perf1", (int)(Mth.clamp(System.nanoTime()-tick, 0, 0x7fffffff))/1000);
        if(!device_enabled) {
          logic_.output_data = 0;
        } else {
          logic_.symbol(".clock", (int)(world.getGameTime() & 0x7fffffffL)); // wraps over in >3years
          logic_.symbol(".time", (int)(world.getDayTime() % 24000));
          logic_.tick();
          if((logic_.rca_output_mask != 0) && (rca_data != RcaSync.CommonRca.EMPTY)) rca_data.server_outputs(logic_.rca_output_data);
        }
      }
      // Output setting
      {
        if(logic_.output_data != last_output_data) {
          for(Direction d:Direction.values()) {
            if((logic_.output_mask & 0xf<<(4*d.ordinal())) == 0) continue;
            final Direction world_dir = ControlBoxBlock.getForwardStateMappedFacing(device_state, d);
            device_block.notifyOutput(device_state, world, device_pos, world_dir);
          }
        }
        if((logic_.output_data != last_output_data) || (logic_.input_data != last_input_data)) world.blockEntityChanged(device_pos);
      }
    } catch(Throwable ex) {
      Auxiliaries.logError("RLC tick exception!" + ex);
      world.removeBlock(getBlockPos(), true);
      return;
    }
    // Elision of signal updates during this tick, including after output setting
    {
      if(logic_.symbols().containsKey("tickrate")) {
        tick_interval_ = Mth.clamp(logic_.symbols().getOrDefault("tickrate", 0), 0, 200);
        if(tick_interval_ == 0) tick_interval_ = TICK_INTERVAL;
      }
      tick_timer_ = tick_interval_;
      final int dl = logic_.symbol(".deadline");
      if((dl>0) && (dl<tick_timer_)) tick_timer_ = dl;
      logic_.intr_redges = 0;
      logic_.intr_fedges = 0;
      if(trace_) logic_.symbol(".perf2", (int)(Mth.clamp(System.nanoTime()-tick, 0, 0x7fffffff))/1000);
    }
  }

  @Override
  public void onServerPacketReceived(CompoundTag nbt)
  { readnbt(getLevel().registryAccess(), nbt); }

  // -------------------------------------------------------------------------------------------

  public boolean getEnabled()
  {
    // @todo: Transitional to prevent breaking setups. ON/OFF state will be "powered".
    return (getBlockState().getValue(ControlBoxBlock.STATE)!=0) || (getBlockState().getValue(ControlBoxBlock.POWERED));
  }

  public void setEnabled(boolean en)
  {
    if(en == getEnabled()) return;
    // @todo: Transitional to prevent breaking setups. ON/OFF state will be "powered".
    getLevel().setBlock(getBlockPos(), getBlockState().setValue(ControlBoxBlock.STATE, en?1:0), 1|2|16);
    getLevel().setBlock(getBlockPos(), getBlockState().setValue(ControlBoxBlock.POWERED, en), 1|2|16);
    if(!en) {
      logic_.clearSymbols();
      final RcaSync.RcaData rca_data = ((logic_.rca_output_mask)==0) ? (RcaSync.CommonRca.EMPTY) : RcaSync.CommonRca.ofPlayer(activating_player_, false);
      if(rca_data != RcaSync.CommonRca.EMPTY) rca_data.server_outputs(0);
    }
  }

  public void setRcaPlayerUUID(@Nullable UUID puid)
  {  activating_player_ = (puid==null) ? (null) : (UUID.fromString(puid.toString())); }

  public String getCode()
  { return logic_.code(); }

  public void setCode(String text)
  { logic_.code(text); }

  public int getOutputSignal(Direction internalSide)
  { return (logic_.output_data >> (4 * internalSide.ordinal())) & 0xf; }

  void scheduleImmediateTick()
  { tick_timer_ = 0; }

  CompoundTag collectSyncData(boolean full)
  {
    final Level world = getLevel();
    final CompoundTag nbt = new CompoundTag();
    nbt.putString("action", "serverdata");
    nbt.putBoolean("enabled", getEnabled());
    nbt.putInt("inputs", logic_.input_mask);
    nbt.putInt("outputs", logic_.output_mask);
    nbt.putInt("ports", (logic_.input_data & logic_.input_mask) | (logic_.output_data & logic_.output_mask));
    if(!logic_.symbols().isEmpty()) {
      final CompoundTag sym_nbt = new CompoundTag();
      logic_.symbols().forEach(sym_nbt::putInt);
      nbt.put("symbols", sym_nbt);
    }
    if(!logic_.valid()) {
      final CompoundTag err_nbt = new CompoundTag();
      logic_.errors().forEach((e,l)->err_nbt.putString(e.toString(), l));
      nbt.put("errors", err_nbt);
    } else {
      nbt.put("errors", new CompoundTag());
    }
    if(!full) return nbt;
    nbt.putBoolean("debug", trace_enabled());
    nbt.putString("code", getCode());
    if(activating_player_ != null) {
      final Player run_player = world.getPlayerByUUID(activating_player_);
      nbt.putString("player", (run_player == null) ? "" : run_player.getScoreboardName());
    }
    return nbt;
  }

  public void signal_update(Direction from_world_side, Direction from_mapped_side)
  {
    if(tick_interval_ > 0) return; // Fixed sample tick interval, RLC not reacting to signal edges.
    final int shift = 4*from_mapped_side.ordinal();
    final int mask = 0xf<<shift;
    if((logic_.input_mask & mask) == 0) return; // no input there
    int signal_intr = mask & (getLevel().getSignal(getBlockPos().relative(from_world_side), from_world_side)<<shift);
    int signal_data = mask & (logic_.input_data);
    if(signal_intr == signal_data) return; // no signal change
    if((signal_intr!=0) && (signal_data==0)) {
      logic_.intr_redges |= mask;
      tick_timer_ = 0;
    } else if(signal_intr==0) {
      logic_.intr_fedges |= mask;
      tick_timer_ = 0;
    }
    // Else no boolean "powered" signal changed. No need to update next tick.
  }

  public void toggle_trace(@Nullable Player player)
  { trace_ = !trace_; if(player!=null) Auxiliaries.playerChatMessage(player, "Trace: " + trace_); }

  public boolean trace_enabled()
  { return trace_; }

  public static final class TestHooks
  {
    private final ControlBoxLogic logic_ = new ControlBoxLogic();

    public boolean setCode(String text)
    { return logic_.code(text); }

    public boolean valid()
    { return logic_.valid(); }

    public Map<Integer, String> errors()
    { return Map.copyOf(logic_.errors()); }

    public int inputMask()
    { return logic_.input_mask; }

    public int outputMask()
    { return logic_.output_mask; }

    public void setInput(Direction side, int value)
    {
      final int shift = 4 * side.ordinal();
      final int mask = 0xf << shift;
      logic_.input_mask |= mask;
      logic_.input_data = (logic_.input_data & ~mask) | ((value & 0xf) << shift);
    }

    public void setSymbol(String key, int value)
    { logic_.symbol(key, value); }

    public int getSymbol(String key)
    { return logic_.symbol(key); }

    public void tick()
    { logic_.tick(); }

    public int output(Direction side)
    { return (logic_.output_data >> (4 * side.ordinal())) & 0xf; }

    public int outputData()
    { return logic_.output_data; }

    public void setRcaInput(int channel, int value)
    {
      final long mask = 0xfL << (channel * 4);
      logic_.rca_input_mask |= mask;
      logic_.rca_input_data = (logic_.rca_input_data & ~mask) | (((long)(value & 0xf)) << (channel * 4));
    }

    public int getRcaOutput(int channel)
    { return (int)((logic_.rca_output_data >> (channel * 4)) & 0xfL); }

    public long rcaOutputData()
    { return logic_.rca_output_data; }

  }

}

File: common/src/main/kotlin/wile/redstonepen/blocks/controlbox/ControlBoxBlockEntity.kt
package wile.redstonepen.blocks.controlbox

import java.util.UUID
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.core.HolderLookup
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.Tag
import net.minecraft.network.chat.Component
import net.minecraft.util.Mth
import net.minecraft.world.Container
import net.minecraft.world.MenuProvider
import net.minecraft.world.Nameable
import net.minecraft.world.SimpleContainer
import net.minecraft.world.entity.player.Inventory
import net.minecraft.world.entity.player.Player
import net.minecraft.world.inventory.AbstractContainerMenu
import net.minecraft.world.inventory.ContainerLevelAccess
import net.minecraft.world.inventory.SimpleContainerData
import net.minecraft.world.level.block.state.BlockState
import wile.redstonepen.blocks.CircuitComponents
import wile.redstonepen.blocks.StandardEntityBlocks
import wile.redstonepen.detail.RcaSync
import wile.redstonepen.net.Networking
import wile.redstonepen.registry.Registries
import wile.redstonepen.util.Auxiliaries

@Suppress("DEPRECATION")
class ControlBoxBlockEntity(pos: BlockPos, state: BlockState) :
    StandardEntityBlocks.StandardBlockEntity(
        requireNotNull(Registries.getBlockEntityTypeOfBlock(state.block)),
        pos,
        state,
    ),
    MenuProvider,
    Nameable,
    Networking.IPacketTileNotifyReceiver {

    companion object {
        const val TICK_INTERVAL = 4
    }

    private val blockInventory: Container = SimpleContainer(1)
    private val logic = ControlBoxLogic()
    private var activatingPlayer: UUID? = null
    private var customName: Component? = null
    private var trace = false
    private var tickTimer = 0
    private var tickInterval = 0
    private var tickErrorMessage: String? = null

    override fun readnbt(hlp: HolderLookup.Provider, nbt: CompoundTag): CompoundTag {
        if (nbt.contains("name", Tag.TAG_STRING.toInt())) {
            customName = Auxiliaries.unserializeTextComponent(nbt.getString("name"), hlp)
        }
        val logicdata =
            if (nbt.contains("logic", Tag.TAG_COMPOUND.toInt())) {
                nbt.getCompound("logic")
            } else {
                CompoundTag()
            }
        logic.code(logicdata.getString("code"))
        logic.inputData = logicdata.getInt("input")
        logic.outputData = logicdata.getInt("output")
        val logicsymbols =
            if (logicdata.contains("symbols", Tag.TAG_COMPOUND.toInt())) {
                logicdata.getCompound("symbols")
            } else {
                CompoundTag()
            }
        logic.clearSymbols()
        logicsymbols.allKeys.forEach { k -> logic.symbol(k, logicsymbols.getInt(k)) }
        activatingPlayer = if (nbt.hasUUID("player")) nbt.getUUID("player") else null
        return nbt
    }

    override fun writenbt(
        hlp: HolderLookup.Provider,
        nbt: CompoundTag,
        syncPacket: Boolean,
    ): CompoundTag {
        customName?.let { nbt.putString("name", Auxiliaries.serializeTextComponent(it, hlp)) }
        val logicdata = CompoundTag()
        logicdata.putString("code", logic.code())
        logicdata.putInt("input", logic.inputData)
        logicdata.putInt("output", logic.outputData)
        val logicsymbols = CompoundTag()
        logic.symbols().forEach { (k, v) -> logicsymbols.putInt(k, v) }
        logicdata.put("symbols", logicsymbols)
        nbt.put("logic", logicdata)
        activatingPlayer?.let { nbt.putUUID("player", it) }
        return nbt
    }

    override fun getName(): Component =
        customName ?: Component.translatable(blockState.block.descriptionId)

    override fun getCustomName(): Component? = customName

    override fun hasCustomName(): Boolean = (customName != null)

    fun setCustomName(name: Component?) {
        customName = name
    }

    override fun getDisplayName(): Component = super.getDisplayName()

    override fun createMenu(id: Int, inventory: Inventory, player: Player): AbstractContainerMenu =
        ControlBoxUiContainer(
            id,
            inventory,
            blockInventory,
            ContainerLevelAccess.create(
                requireNotNull(level) {
                    "BlockEntity must be attached to a level before createMenu is called"
                },
                worldPosition,
            ),
            SimpleContainerData(1),
        )

    override fun tick() {
        if (tickErrorMessage != null || --tickTimer > 0) return
        tickTimer = if (tickInterval > 0) tickInterval else TICK_INTERVAL
        val tick = System.nanoTime()
        val world = getLevel() ?: return
        val deviceState = blockState
        val devicePos = blockPos
        val deviceEnabled =
            (deviceState.getValue(CircuitComponents.DirectedComponentBlock.STATE) > 0) ||
                (deviceState.getValue(CircuitComponents.DirectedComponentBlock.POWERED))
        if (deviceState.block !is ControlBoxBlock) return
        val deviceBlock = deviceState.block as ControlBoxBlock
        val lastOutputData = logic.outputData
        val lastInputData = logic.inputData
        val rcaData =
            if ((logic.rcaInputMask or logic.rcaOutputMask) == 0L) {
                RcaSync.CommonRca.EMPTY
            } else {
                RcaSync.CommonRca.ofPlayer(activatingPlayer, false)
            }
        try {
            run {
                logic.inputData = 0
                for (d in Direction.entries) {
                    val worldDir =
                        CircuitComponents.DirectedComponentBlock.getForwardStateMappedFacing(
                            deviceState,
                            d,
                        )
                    if (deviceEnabled) {
                        val portName = PortNames.ALL[d.ordinal]
                        if (logic.usesSymbol("${portName}.co")) {
                            val targetPos = devicePos.relative(worldDir)
                            val targetState = world.getBlockState(targetPos)
                            logic.symbol(
                                "${portName}.co",
                                if (targetState.hasAnalogOutputSignal()) {
                                    targetState.getAnalogOutputSignal(world, targetPos)
                                } else {
                                    0
                                },
                            )
                        }
                    }
                    if ((logic.outputMask and (0xf shl (4 * d.ordinal))) == 0) {
                        val p = world.getSignal(devicePos.relative(worldDir), worldDir)
                        logic.inputData = logic.inputData or ((p and 0xf) shl (4 * d.ordinal))
                    }
                }
                if (logic.rcaInputMask != 0L && rcaData != RcaSync.CommonRca.EMPTY) {
                    logic.rcaInputData = rcaData.client_inputs() and logic.rcaInputMask
                }
            }
            run {
                if (trace) {
                    logic.symbol(
                        ".perf1",
                        (Mth.clamp(System.nanoTime() - tick, 0L, 0x7fffffffL) / 1000L).toInt(),
                    )
                }
                if (!deviceEnabled) {
                    logic.outputData = 0
                } else {
                    logic.symbol(".clock", (world.gameTime and 0x7fffffffL).toInt())
                    logic.symbol(".time", (world.dayTime % 24000L).toInt())
                    logic.tick()
                    if (logic.rcaOutputMask != 0L && rcaData != RcaSync.CommonRca.EMPTY) {
                        rcaData.server_outputs(logic.rcaOutputData)
                    }
                }
            }
            run {
                if (logic.outputData != lastOutputData) {
                    for (d in Direction.entries) {
                        if ((logic.outputMask and (0xf shl (4 * d.ordinal))) == 0) continue
                        val worldDir =
                            CircuitComponents.DirectedComponentBlock.getForwardStateMappedFacing(
                                deviceState,
                                d,
                            )
                        deviceBlock.notifyOutput(deviceState, world, devicePos, worldDir)
                    }
                }
                if (logic.outputData != lastOutputData || logic.inputData != lastInputData) {
                    world.blockEntityChanged(devicePos)
                }
            }
        } catch (ex: Throwable) {
            val msg = "ControlBox halted: ${ex.message ?: ex.javaClass.simpleName}"
            Auxiliaries.logError(msg)
            tickErrorMessage = msg
            activatingPlayer?.let { uuid ->
                world.getPlayerByUUID(uuid)?.let { Auxiliaries.playerChatMessage(it, msg) }
            }
            setEnabled(false)
            setChanged()
            return
        }
        run {
            if (logic.symbols().containsKey("tickrate")) {
                tickInterval = Mth.clamp(logic.symbols().getOrDefault("tickrate", 0), 0, 200)
                if (tickInterval == 0) tickInterval = TICK_INTERVAL
            }
            tickTimer = tickInterval
            val dl = logic.symbol(".deadline")
            if (dl > 0 && dl < tickTimer) tickTimer = dl
            logic.risingEdgeInterrupts = 0
            logic.fallingEdgeInterrupts = 0
            if (trace) {
                logic.symbol(
                    ".perf2",
                    (Mth.clamp(System.nanoTime() - tick, 0L, 0x7fffffffL) / 1000L).toInt(),
                )
            }
        }
    }

    override fun onServerPacketReceived(nbt: CompoundTag) {
        val world = getLevel() ?: return
        readnbt(world.registryAccess(), nbt)
    }

    fun getEnabled(): Boolean =
        (blockState.getValue(CircuitComponents.DirectedComponentBlock.STATE) != 0) ||
            (blockState.getValue(CircuitComponents.DirectedComponentBlock.POWERED))

    fun setEnabled(en: Boolean) {
        if (en == getEnabled()) return
        val world = getLevel() ?: return
        world.setBlock(
            blockPos,
            blockState.setValue(CircuitComponents.DirectedComponentBlock.STATE, if (en) 1 else 0),
            1 or 2 or 16,
        )
        world.setBlock(
            blockPos,
            blockState.setValue(CircuitComponents.DirectedComponentBlock.POWERED, en),
            1 or 2 or 16,
        )
        if (!en) {
            logic.clearSymbols()
            val rcaData =
                if (logic.rcaOutputMask == 0L) {
                    RcaSync.CommonRca.EMPTY
                } else {
                    RcaSync.CommonRca.ofPlayer(activatingPlayer, false)
                }
            if (rcaData != RcaSync.CommonRca.EMPTY) rcaData.server_outputs(0L)
        }
    }

    fun setRcaPlayerUUID(puid: UUID?) {
        activatingPlayer = if (puid == null) null else UUID.fromString(puid.toString())
    }

    fun getCode(): String = logic.code()

    fun setCode(text: String) {
        tickErrorMessage = null
        logic.code(text)
    }

    fun getOutputSignal(internalSide: Direction): Int =
        (logic.outputData shr (4 * internalSide.ordinal)) and 0xf

    internal fun scheduleImmediateTick() {
        tickTimer = 0
    }

    internal fun collectSyncData(full: Boolean): CompoundTag {
        val nbt = CompoundTag()
        nbt.putString("action", "serverdata")
        nbt.putBoolean("enabled", getEnabled())
        nbt.putInt("inputs", logic.inputMask)
        nbt.putInt("outputs", logic.outputMask)
        nbt.putInt(
            "ports",
            (logic.inputData and logic.inputMask) or (logic.outputData and logic.outputMask),
        )
        if (logic.symbols().isNotEmpty()) {
            val symNbt = CompoundTag()
            logic.symbols().forEach { (k, v) -> symNbt.putInt(k, v) }
            nbt.put("symbols", symNbt)
        }
        if (!logic.valid()) {
            val errNbt = CompoundTag()
            logic.errors().forEach { (e, l) -> errNbt.putString(e.toString(), l) }
            nbt.put("errors", errNbt)
        } else {
            nbt.put("errors", CompoundTag())
        }
        tickErrorMessage?.let { nbt.putString("runtimeError", it) }
        if (!full) return nbt
        nbt.putBoolean("debug", trace_enabled())
        nbt.putString("code", getCode())
        activatingPlayer?.let { pid ->
            val runPlayer = getLevel()?.getPlayerByUUID(pid)
            nbt.putString("player", runPlayer?.scoreboardName ?: "")
        }
        return nbt
    }

    fun signal_update(fromWorldSide: Direction, fromMappedSide: Direction) {
        if (tickInterval > 0) return
        val shift = 4 * fromMappedSide.ordinal
        val mask = 0xf shl shift
        if ((logic.inputMask and mask) == 0) return
        val world = getLevel() ?: return
        val signalIntr =
            mask and (world.getSignal(blockPos.relative(fromWorldSide), fromWorldSide) shl shift)
        val signalData = mask and logic.inputData
        if (signalIntr == signalData) return
        if (signalIntr != 0 && signalData == 0) {
            logic.risingEdgeInterrupts = logic.risingEdgeInterrupts or mask
            tickTimer = 0
        } else if (signalIntr == 0) {
            logic.fallingEdgeInterrupts = logic.fallingEdgeInterrupts or mask
            tickTimer = 0
        }
    }

    fun toggle_trace(player: Player?) {
        trace = !trace
        if (player != null) Auxiliaries.playerChatMessage(player, "Trace: $trace")
    }

    fun trace_enabled(): Boolean = trace

    class TestHooks {
        private val logic = ControlBoxLogic()
        private var tickErrorMessage: String? = null

        fun setCode(text: String): Boolean {
            tickErrorMessage = null
            return logic.code(text)
        }

        fun valid(): Boolean = logic.valid()

        fun errors(): Map<Int, String> = HashMap(logic.errors())

        fun inputMask(): Int = logic.inputMask

        fun outputMask(): Int = logic.outputMask

        fun setInput(side: Direction, value: Int) {
            val shift = 4 * side.ordinal
            val mask = 0xf shl shift
            logic.inputMask = logic.inputMask or mask
            logic.inputData = (logic.inputData and mask.inv()) or ((value and 0xf) shl shift)
        }

        fun setSymbol(key: String, value: Int) {
            logic.symbol(key, value)
        }

        fun getSymbol(key: String): Int = logic.symbol(key)

        fun tick() {
            logic.tick()
        }

        fun simulateTick() {
            if (tickErrorMessage != null) return
            try {
                logic.tick()
            } catch (ex: Throwable) {
                tickErrorMessage = ex.message ?: ex.javaClass.simpleName
            }
        }

        fun tickErrorMessage(): String? = tickErrorMessage

        fun injectTickError(message: String) {
            tickErrorMessage = message
        }

        fun output(side: Direction): Int = (logic.outputData shr (4 * side.ordinal)) and 0xf

        fun outputData(): Int = logic.outputData

        fun setRcaInput(channel: Int, value: Int) {
            val mask = 0xfL shl (channel * 4)
            logic.rcaInputMask = logic.rcaInputMask or mask
            logic.rcaInputData =
                (logic.rcaInputData and mask.inv()) or ((value.toLong() and 0xfL) shl (channel * 4))
        }

        fun getRcaOutput(channel: Int): Int =
            ((logic.rcaOutputData shr (channel * 4)) and 0xfL).toInt()

        fun rcaOutputData(): Long = logic.rcaOutputData
    }
}
